### PR TITLE
Document Git workflow and guidelines

### PR DESCRIPTION
**Pull request by ranweiler**

### CODE REVIEW

**Comment by nealmcb** [comment]
I'm impressed, Joe - very thoughtful.  Thanks!
New stuff for me were

- The seven rules of a great Git commit message, and
- To support bisecting, do not merge WIP commits that break the build. On topic branches, squash commits as needed before merging.

Makes great sense.

**Comment by nealmcb** [comment]
Do you want to say something about preferring rebase to merges?

**Comment by ranweiler** [comment]
@nealmcb, hmm, I'm fine with whatever, here. I forgot that we can now have GitHub do the rebase for us, though using the button means it won't create a merge commit. Any thoughts from others?

**Comment by dmzimmerman** [comment]
Personally, I prefer rebases to merges because I find merge commits to be mostly superfluous and information-free. But I'm fine with whatever is decided.

**Comment by ranweiler** [comment]
With merging, when using a GitHub (and similar) PR-based workflow, one nice thing about the generated merge commit is that it has a reference back to the PR, which GitHub knows how to turn into a hyperlink in its web interface. Other than that, and issues that can sometimes arise with custom CI systems, I agree, you don't get much, and I aesthetically prefer rebasing + no merge commit.

Oh, actually: I assume an automated, GitHub-driven rebase would invalidate the commit signing.

**Comment by ranweiler** [comment]
@dmzimmerman and @nealmcb, I'm going to try an experiment on a private GitHub branch on my own account and see how the rebase merges interact with signing. Will report back shortly.

**Comment by ranweiler** [comment]
Okay, yes, as I thought, you lose the signature if a non-trivial rebase happens. So, we should merge, since then only the merge commit will be unsigned. We can have it all if we eschew button-press merging, but I'm not sure that's necessary.

**Comment by nealmcb** [comment]
@ranweiler I'm used to just doing the `git rebase master` on my workstation.  What would it mean to "eschew button-press merging"? If you try it, does it warn you you're going to lose a signature?

**Comment by ranweiler** [comment]
@nealmcb ah, I see. So, I'm thinking about the Git*Hub* side of thing: in a PR, we have that big green "Merge pull request" button. GitHub now lets you choose a few different strategies (by clicking on the green down arrow on the button). That's what I'm thinking about right now. If you choose the "Rebase and merge" strategy, GitHub will rebase your branch on master _for_ you, and then merge the PR's topic branch (without creating a merge commit). The problem with this for us is that you lose the signature on the commits being merged-via-rebase: since GitHub doesn't have your private key, it can't re-sign them, which would be necessary after rebasing, since the commit hashes change.

On the other hand, if you use the standard "Merge pull request" strategy, GitHub doesn't do any rebasing for you, and it _does_ create a merge commit. This preserves the hashes, thus the signatures, so only the newly-created (by GitHub) merge commit is unsigned.

We could combine the benefits by requiring PR submitters to locally rebase onto master, as you suggest, and manually re-sign any commit hashes modified by their local rebase. Then the GitHub button-based "Rebase and merge" strategy would (1) not alter any commits, thus preserving signatures, (2) not create a (naturally unsigned) merge commit.

**Comment by ranweiler** [comment]
FWIW, I'm perfectly happy to do what you mention (manually rebase locally), then use the "Rebase and merge" strategy from within GitHub.

**Comment by ranweiler** [comment]
@dmz, @kiniry, any thoughts on the above?

**Comment by ranweiler** [comment]
You know what: even that doesn't work. The "Rebase and merge" strategy always breaks signatures, even if it didn't need to do a nontrivial rebase on master. AFAICT, to preserve signatures when merging via GitHub, we must use the non-rebasing, "Merge pull request" strategy, which results in an unsigned merge commit. 

Independent of that, we may or may not choose to ask that PR submitters locally rebase (and re-sign) on master (like @nealmcb described) before having their reviewer merge.

**Comment by nealmcb** [comment]
How much checking is or could be done to leverage the pgp signatures? Without good checking, anyone can create a key pair, associate any name with it, and make a signature. So does this extra effort actually buy us much?  E.g. will Travis CI ensure that we only accept code from people we've approved (or allow "countersigning" of some sort)?

**Comment by dmzimmerman** [comment]
I'm not sure how much of that can be done automatically, Neal, but we're only accepting code from people we've approved anyway (by virtue of not giving external parties push access to the repo). So in a sense I suppose it doesn't buy us much.

**Comment by ranweiler** [comment]
@nealmcb we could make `master` a "protected" branch, then use a "status check" (i.e. CI integration) to ensure that one cannot merge to master without a successful status check.

As @dmzimmerman says, we're already leaning on GitHub for a lot. A benefit of using signatures is that it gives us an out-of-band mechanism for authentication and non-repudiation. We don't have a formal threat model for our dev process, though, so these questions are hard to answer well.

**Comment by ranweiler** [comment]
IMO, we should just use signatures and normal merges, and not require rebasing on `master`. This gets us a lot, is conservative without a lot of extra workflow shenanigans, and we can use GitHub features to automate some checks. It also gives us the links to PRs in the merge commits, as I mentioned earlier.

**Comment by nealmcb** [comment]
So I guess I should add nealmcb@freeandfair.us to [my PGP key](http://pgp.es.net:11371/pks/lookup?op=vindex&search=0x73B18E7E5736E0ED)?

It seems like when I re-signed my code, it stopped using my freeandfair email address as "Author" - not sure why. But it kept the original dates in the past.

And it doesn't seem like there is any checking of pgp key vs committer id.  See

```
commit d652ed10205c597eedeb342ee05ed36d67600ec1
gpg: Signature made Thu 13 Jul 2017 09:51:30 AM MDT using RSA key ID 5736E0ED
gpg: Good signature from "Neal McBurnett <neal@mcburnett.org>"
gpg:                 aka "Neal McBurnett <nealmcb@bell-labs.com>"
gpg:                 aka "Neal McBurnett <neal.mcburnett@att.com>"
gpg:                 aka "Neal McBurnett <neal@dr.att.com> expires 1998/04/01"
Merge: bc66ca0 e169248
Author: Neal McBurnett <neal@mcburnett.org>
Date:   Tue Jul 11 16:43:06 2017 -0600

    Merge pull request #20 from FreeAndFair/fix-readme-links
    
    Fix README links to documentation

commit e1692485ed4c3d11f2b3f3029d5fe1e1d2dd3692
gpg: Signature made Thu 13 Jul 2017 09:51:27 AM MDT using RSA key ID 5736E0ED
gpg: Good signature from "Neal McBurnett <neal@mcburnett.org>"
gpg:                 aka "Neal McBurnett <nealmcb@bell-labs.com>"
gpg:                 aka "Neal McBurnett <neal.mcburnett@att.com>"
gpg:                 aka "Neal McBurnett <neal@dr.att.com> expires 1998/04/01"
Author: Joe Ranweiler <josephkennethranweiler@gmail.com>
Date:   Tue Jul 11 15:28:54 2017 -0700

    Fix README links to documentation.
```

And I guess we'll need to achieve trust in each other's keys, in order to pull code?

I'm still a bit dubious about the cost/benefit ratio of signing commits, mostly fearing that more little gotchas like the issues around rebasing will keep cropping up.

**Comment by dmzimmerman** [comment]
There is a check between committer ID and signer ID, but only on GitHub’s website I think (it says “verified” vs. “unverified” for signatures that come from addresses that are not verified with GitHub when you look at the list). I’m not sure how to get that data from the command line.

**Comment by nealmcb** [comment]
The **right** think to work towards would be secure updates via [TUF - The Update Framework](https://theupdateframework.github.io/)!
Real signatures by keys meant to sign artifacts and update them and recover from failure, and so much more.

**Comment by ranweiler** [comment]
@nealmcb, try the following:
1. Return your local `glossary2` branch to its original state, i.e. _before_ you tried to rebase or ran `git filter-branch ...`. You may need to `git reset --hard origin/glossary2`, unless you made another local copy.
1. To rebase onto `master`, signing _only the new commits_ in `glossary2`, try: `git rebase -S master` (note the `-S`). This will _not_ sign older commits that you've authored that are _not_ part of the `glossary2` branch. It will only sign the `glossary2` commits that are being replayed onto `master`.

Suppose you had a branch `other` which was already based on the latest `master`— in other words, doing `git rebase master` wouldn't do anything. You can still easily sign _only_ your new commits on the `other` branch by add the `-f` flag, so, the above would become: `git rebase -f -S master`. We can thus avoid using `git filter-branch` except in more complicated scenarios. Moving forward, we shouldn't need it at all.

**Comment by nealmcb** [comment]
My git-rebase has no -S option, I'm afraid.... `git version 1.9.1` (Trusty). Do you know when it was added?

**Comment by ranweiler** [comment]
@nealmcb hmm, looks like it first appears in the Git docs for 2.0.0.

**Comment by nealmcb** [comment]
Thanks to @ranweiler's awesome, indefatigable helpfulness, I've got https://launchpad.net/~git-core/+archive/ubuntu/ppa and git 1:2.13.0 installed now and working like a champ :)

I think my newly-signed glossary2 branch https://github.com/FreeAndFair/ColoradoRLA/pull/22 is ready to pull now.

File: docs/25_developer.md [hunk 1]
```diff
@@ -171,6 +171,55 @@ level, complete with versioning information. Note that we prefer that
...
+  [seven rules](https://chris.beams.io/posts/git-commit/#seven-rules),
+  with the amendment that on this project, we have adopted the
+  convention of ending the subject line with a period.
+- Keep your topic branches small to facilitate review.
+- Before merging someone else's PR, make sure other reviewers'
+  comments are resolved, and that the PR author considers the PR ready
+  to merge.
+- For security-sensitive code, ensure your changes have received an
+  in-depth review, preferably from multiple reviewers.
+- Configure Git so that your commits
```
Comment (dmzimmerman) [inline docs/25_developer.md]: This line and the next line are both quite short, and should probably be combined.

File: docs/25_developer.md [hunk 2]
```diff
@@ -171,6 +171,55 @@ level, complete with versioning information. Note that we prefer that
...
+  are
+  [signed](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work).
```
Comment (dmzimmerman) [inline docs/25_developer.md]: We should perhaps add something explicit here about how to perform merges of pull requests in the presence of signatures so that the signatures are preserved.

File: docs/25_developer.md (hunk 1 above)
Comment (ranweiler) [inline docs/25_developer.md]: Do you mean the item about linters? I have them separate because they are about different things (signatures vs. not committing noise). With that in mind, would you still suggest I merge them?

File: docs/25_developer.md (hunk 2 above)
Comment (ranweiler) [inline docs/25_developer.md]: Will do.

File: docs/25_developer.md (hunk 1 above)
Comment (dmzimmerman) [inline docs/25_developer.md]: No, I was imprecise. I was looking at the two lines of the file, "- Configure Git so that your commits" and "are", and thinking that "are" should be moved to the previous line (in the spirit of avoiding unnecessary whitespace). So I was literally talking about "lines" as presented in GitHub's review interface rather than items in Markdown.

File: docs/25_developer.md (hunk 2 above)
Comment (ranweiler) [inline docs/25_developer.md]: Done.

File: docs/25_developer.md (hunk 1 above)
Comment (ranweiler) [inline docs/25_developer.md]: Oh, I see! Agreed, that's a weird artifact of Emacs' automatic paragraph formatting. Fixing.

Comment (ranweiler) [inline docs/25_developer.md]: Done.

Comment (dmzimmerman) [inline docs/25_developer.md]: I have also had that issue with Emacs' paragraph formatting, and it always strikes me as bizarre when it happens. I feel like that algorithm must be fundamentally broken in some way.

**Comment by nealmcb** [comment]
I just ran across some info relevant to a thought I mentioned above. The Update Framework is aimed at securing updates of packaged software systems. The same lab is working on [in-toto](https://in-toto.github.io/), which is intended to be more relevant for securing the software supply chain.
Some day it would be interesting to see if that would be useful for the sort of things we're talking about here.
